### PR TITLE
INTYGFV-16453: Implementation for generation of pdf

### PR DESCRIPTION
**Pull request by mikaelaromild1**

### CODE REVIEW

**Comment by mikaelaromild1** [comment]
@mww-sopra-steria Vill gärna diskutera denna jira med dig på måndag om du har möjlighet då det råder osäkerheter kring några punkter, bland annat:

- vilka exceptions vi ska kasta i CertificatePdfGenerator (för att komma vidare satte jag bara IllegalStateException på alla tills vidare)
- vart det är bäst att sätta pdfen som readonly eftersom möjligheten att testa pdfen försvinner när man plattar den
- om den specifika koden för FK7211 ska brytas ut till egen klass eller om det bör göras i ett senare skede

File: certificate-service/pdfbox-generator/src/main/java/se/inera/intyg/certificateservice/pdfboxgenerator/FK7211PdfGenerator.java [hunk 1]
```diff
@@ -0,0 +1,224 @@
...
+      "Detta är en utskrift av ett elektroniskt intyg. "
+          + "Intyget har signerats elektroniskt av intygsutfärdaren.";
+
+
+  private final PDDocument pdDocument;
+  private final PDAcroForm acroForm;
+
+  public FK7211PdfGenerator(Certificate certificate) throws IOException {
+    String template = certificate.certificateModel().pdfTemplatePath();
+    InputStream inputStream = getClass().getClassLoader().getResourceAsStream(template);
```
Comment (mww-sopra-steria) [inline certificate-service/pdfbox-generator/src/main/java/se/inera/intyg/certificateservice/pdfboxgenerator/FK7211PdfGenerator.java:67]: I think we need to wrap the inputstream code in a try-catch so we make sure that it is closed (maybe can even use try with resource here)

File: certificate-service/app/build.gradle [hunk 2]
```diff
@@ -15,6 +15,7 @@ dependencies {
     implementation 'org.springframework.boot:spring-boot-starter-actuator'
     implementation 'org.springframework.boot:spring-boot-starter-data-jpa'
     implementation 'org.springframework.boot:spring-boot-starter-web'
+    implementation project(path: ':pdfbox-generator')
```
Comment (mww-sopra-steria) [inline certificate-service/app/build.gradle:18]: Move this to the top of dependencies. Makes it easier to see what dependencies we have between gradle-modules.

File: certificate-service/app/src/main/java/se/inera/intyg/certificateservice/application/certificate/CertificateInternalApiController.java [hunk 3]
```diff
@@ -14,10 +16,18 @@
...
+  private final GetCertificateInternalPdfService getCertificateInternalPdfService;
 
   @PostMapping("/{certificateId}/xml")
   GetCertificateInternalXmlResponse getCertificateXml(
       @PathVariable("certificateId") String certificateId) {
     return getCertificateInternalXmlService.get(certificateId);
   }
+
+  @PostMapping("/{certificateId}/pdf")
+  GetCertificateInternalPdfResponse getCertificatePdf(
```
Comment (mww-sopra-steria) [inline certificate-service/app/src/main/java/se/inera/intyg/certificateservice/application/certificate/CertificateInternalApiController.java:28]: We should not expose pdf-printing ni the internal api. Should be removed together with any "internal"-implementation classes

File: certificate-service/app/src/main/java/se/inera/intyg/certificateservice/application/certificate/service/validation/GetCertificatePdfRequestValidator.java [hunk 4]
```diff
@@ -0,0 +1,21 @@
...
+import static se.inera.intyg.certificateservice.application.common.validator.ValidationUtil.validateUnitExtended;
+import static se.inera.intyg.certificateservice.application.common.validator.ValidationUtil.validateUser;
+
+import org.springframework.stereotype.Component;
+import se.inera.intyg.certificateservice.application.certificate.dto.GetCertificatePdfRequest;
+
+@Component
+public class GetCertificatePdfRequestValidator {
+
+  public void validate(GetCertificatePdfRequest getCertificatePdfRequest, String certificateId) {
```
Comment (mww-sopra-steria) [inline certificate-service/app/src/main/java/se/inera/intyg/certificateservice/application/certificate/service/validation/GetCertificatePdfRequestValidator.java:14]: Should we validate that additionalInfo is included or is it optional?

File: certificate-service/domain/src/main/java/se/inera/intyg/certificateservice/domain/action/model/CertificateActionPrint.java [hunk 5]
```diff
@@ -0,0 +1,29 @@
...
+import java.util.Optional;
+import lombok.AccessLevel;
+import lombok.Builder;
+import lombok.Getter;
+import se.inera.intyg.certificateservice.domain.certificate.model.Certificate;
+import se.inera.intyg.certificateservice.domain.certificatemodel.model.CertificateActionSpecification;
+
+@Builder
+@Getter(AccessLevel.NONE)
+public class CertificateActionPrint implements CertificateAction {
```
Comment (mww-sopra-steria) [inline certificate-service/domain/src/main/java/se/inera/intyg/certificateservice/domain/action/model/CertificateActionPrint.java:13]: Need to implement name and description for the action. See for example CertificateActionDelete.

**Review by mww-sopra-steria — CHANGES_REQUESTED** [review]
Nice work and well structured code!

See comment and let me know if you have any questions or would like to discuss.

I will run some tests as well, but haven't had time yet, but like to submit the feedback directly so you don't have to wait for my tests.

File: certificate-service/app/src/main/java/se/inera/intyg/certificateservice/application/certificate/service/validation/GetCertificatePdfRequestValidator.java (hunk 4 above)
Comment (mikaelaromild1) [reply certificate-service/app/src/main/java/se/inera/intyg/certificateservice/application/certificate/service/validation/GetCertificatePdfRequestValidator.java:14]: Decided to make it non optional and added validation

File: certificate-service/app/src/main/java/se/inera/intyg/certificateservice/infrastructure/configuration/AppConfig.java [hunk 6]
```diff
@@ -127,4 +130,19 @@ public SendCertificateDomainService sendCertificateDomainService(
...
         certificateEventDomainService);
   }
+
+  @Bean
+  public GetCertificatePdfDomainService getCertificatePdfDomainService(
+      CertificateRepository certificateRepository, PdfGenerator pdfGenerator,
+      CertificateEventDomainService certificateEventDomainService) {
+    return new GetCertificatePdfDomainService(certificateRepository, pdfGenerator,
+        certificateEventDomainService);
+
```
Comment (mww-sopra-steria) [inline certificate-service/app/src/main/java/se/inera/intyg/certificateservice/infrastructure/configuration/AppConfig.java:140]: Unnecessary blank lines

File: certificate-service/app/src/test/java/se/inera/intyg/certificateservice/application/certificate/CertificateInternalApiControllerTest.java [hunk 7]
```diff
@@ -15,6 +15,9 @@
 class CertificateInternalApiControllerTest {
 
   private static final String CERTIFICATE_ID = "certificateId";
+  private static final String FILE_NAME = "fileName";
```
Comment (mww-sopra-steria) [inline certificate-service/app/src/test/java/se/inera/intyg/certificateservice/application/certificate/CertificateInternalApiControllerTest.java:18]: Remove constants that aren't needed.

**Review by mww-sopra-steria — APPROVED** [review]
Looks great! See two comments before merging.

**Pull request merged**